### PR TITLE
RELATED: RAIL-4418 Add first version of attribute elements caching

### PR DESCRIPTION
**Pull request by no23reason**
This only caches the first `query` results, the `next` and `goTo` are not cached yet. Also queries with the `filter` set are not cached either.

JIRA: RAIL-4418

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/9855/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/11751/ : SUCCESS in 20m 56s

**Review by kandl — APPROVED** [review]

File: libs/sdk-backend-base/src/cachingBackend/index.ts [hunk 1]
```diff
@@ -433,6 +444,113 @@ function refMatchesMdObject(ref: ObjRef, mdObject: IMetadataObject, type?: Objec
...
+    settings: {
+        limit?: number;
+        offset?: number;
+        options?: IElementsQueryOptions;
+        attributeFilters?: IElementsQueryAttributeFilter[];
+        dateFilters?: IRelativeDateFilter[];
+        measures?: IMeasure[];
+    },
+): string {
+    return `${objRefToString(ref)}__${stringify(settings)}`;
```
[the remark this answers is not in the document]
Comment (no23reason) [reply libs/sdk-backend-base/src/cachingBackend/index.ts]: Good point, updated :)

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2998

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2804/ : SUCCESS in 2m 15s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3107/ : SUCCESS in 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1949/ : SUCCESS in 35m 38s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/11761/ : SUCCESS in 27m 23s

**Pull request merged**